File: src/compiler/semanticVerifier.java
import symbols.*;

import java.util.function.Function;

import org.antlr.v4.runtime.ParserRuleContext;

@SuppressWarnings("CheckReturnValue")
public class semanticVerifier extends pdrawBaseVisitor<IType> {

   private SymbolTable symbolTable = new SymbolTable();

   public IType switchFunctionType(String name){
      IType type = null;
		switch (name) {
			case "pen":
				type = new Pen();
				break;
			case "real":
				type = new Real();
				break;
			case "int":
				type = new IntegerType();
				break;
			case "string":
				type = new StringType();
				break;
			case "point":
				type = new Point();
				break;
         case "bool":
            type = new BooleanType();
            break;
      }

      return type;
   }

   @Override
	public IType visitProgram(pdrawParser.ProgramContext ctx) {
		visitChildren(ctx);
      return null;
	}

   @Override public IType visitStatement(pdrawParser.StatementContext ctx) {
      visitChildren(ctx);
      return null;
   }

   @Override public IType visitCompound(pdrawParser.CompoundContext ctx) {
      visitChildren(ctx);
      return null;
   }

   @Override public IType visitBlock(pdrawParser.BlockContext ctx) {
      symbolTable.enterScope();
      visitChildren(ctx);
      symbolTable.exitScope();
      return null;
   }

   @Override public IType visitDefine(pdrawParser.DefineContext ctx) {
      visitChildren(ctx);
      return null;
   }

   @Override public IType visitPenDefinition(pdrawParser.PenDefinitionContext ctx) {
      String id = ctx.ID().getText();
      IType type = new PenType();
		Symbol penType = new Symbol(id, type);
		symbolTable.addSymbol(penType);
		return type;
   }

   @Override public IType visitPropertyDefinition(pdrawParser.PropertyDefinitionContext ctx) {
      String property = ctx.Property().getText();
		IType exprType = visit(ctx.expression());

		switch (property) {
			case "color":
				if (exprType.getType() != Type.INTEGER) {
					ErrorHandler.error("Color value must be a integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "pressure":
				if (!isNumericType(exprType.getType())) {
					ErrorHandler.error("Pressure value must be a real or integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "thickness":
				if (exprType.getType() != Type.INTEGER) {
					ErrorHandler.error("Thickness value must be a integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;	
			case "orientation":
				if (!isNumericType(exprType.getType())) {
					ErrorHandler.error("Orientation value must be a real or integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "position":
				if (exprType.getType() != Type.POINT) {
					ErrorHandler.error("Position value must be a point value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
		
			default:
				ErrorHandler.error("Invalid property " + property, ctx, symbolTable.getCurrentFunction());
				break;
		}
		return exprType;
	}

   @Override public IType visitCanvasDefinition(pdrawParser.CanvasDefinitionContext ctx) {
      String id = ctx.ID().getText();
      IType title = visit(ctx.expression(0));
      IType dimension = visit(ctx.expression(1));

      if (title.getType() != Type.STRING) {
         ErrorHandler.error("Canvas title must be a string value.", ctx, symbolTable.getCurrentFunction());
      }

      if (dimension.getType() != Type.POINT) {
         ErrorHandler.error("Canvas dimension must be a point value.", ctx, symbolTable.getCurrentFunction());
      }

      symbolTable.addSymbol(new Symbol(id, new Canvas()));

      return null;
   }

   @Override public IType visitDeclaration(pdrawParser.DeclarationContext ctx) {
      IType type = switchFunctionType(ctx.type.getText());

      for (pdrawParser.Declaration_elementContext element : ctx.declaration_element()) {
         String id = element.ID() != null ? element.ID().getText() : element.assign().ID().getText();
         if (symbolTable.getSymbol(id) != null) {
            ErrorHandler.error("Variable with id " + id + " already exists.", ctx, symbolTable.getCurrentFunction());
         } else {
            Symbol variable = new Symbol(id, type, ctx.start.getLine(), ctx.start.getCharPositionInLine(), ctx.stop.getLine(), ctx.stop.getCharPositionInLine());
            symbolTable.addSymbol(variable);
         }

         visit(element);
      }
		return type;
   }

   @Override public IType visitDeclaration_element(pdrawParser.Declaration_elementContext ctx) {
      if (ctx.assign() != null)
         visit(ctx.assign());

      return null;
   }

   @Override public IType visitWhile(pdrawParser.WhileContext ctx) {
      if (visit(ctx.expression()).getType() != Type.BOOLEAN) {
         ErrorHandler.error("While condition must be a boolean value.", ctx, symbolTable.getCurrentFunction());
      }
      return null;
   }

   @Override public IType visitUntil(pdrawParser.UntilContext ctx) {
      if (visit(ctx.expression()).getType() != Type.BOOLEAN) {
         ErrorHandler.error("Until condition must be a boolean value.", ctx, symbolTable.getCurrentFunction());
      }
      return null;
   }

   @Override public IType visitFor(pdrawParser.ForContext ctx) {
      // check if the loop range is an integer
      if (visit(ctx.expression(0)).getType() != Type.BOOLEAN) {
         ErrorHandler.error("Loop range must be a boolean value.", ctx, symbolTable.getCurrentFunction());
      }
      return null;
   }

   @Override public IType visitIf(pdrawParser.IfContext ctx) {
      if (visit(ctx.expression()).getType() != Type.BOOLEAN) {
         ErrorHandler.error("If condition must be a boolean value.", ctx, symbolTable.getCurrentFunction());
      }
      return null;
   }

   @Override public IType visitExecution(pdrawParser.ExecutionContext ctx) {
      String id = ctx.ID().getText();
      Symbol s = symbolTable.getSymbol(id);
      if (s == null) {
         ErrorHandler.error(id + " was not declared.", ctx, symbolTable.getCurrentFunction());
      }
      if (s.getType() != Type.PEN) {
         ErrorHandler.error("Execution must be applied to pen values.", ctx, symbolTable.getCurrentFunction());
      }

      IType exprType = visit(ctx.expression());
      if (exprType.getType() != Type.STRING) {
         ErrorHandler.error("Execution filename must be a string value.", ctx, symbolTable.getCurrentFunction());
      }

      return null;
   }

   @Override public IType visitPause(pdrawParser.PauseContext ctx) {
      IType exprType = visit(ctx.expression());
		if (exprType.getType() != Type.INTEGER) {
			ErrorHandler.error("Pause time must be an integer value.", ctx, symbolTable.getCurrentFunction());
		}
		return exprType;
   }

   @Override public IType visitStdout(pdrawParser.StdoutContext ctx) {
      IType exprType = visit(ctx.expression());
		if (exprType.getType() != Type.STRING && exprType.getType() != Type.PEN && exprType.getType() != Type.INTEGER && exprType.getType() != Type.REAL) {
			ErrorHandler.error("Stdout value must be convertible to string.", ctx, symbolTable.getCurrentFunction());
		}
		return null;
   }

   @Override public IType visitExprToString(pdrawParser.ExprToStringContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
		if (!isConvertibleToString(exprType)) {
			ErrorHandler.error("Conversion to string can only be applied to real, integer, or string values.", ctx, symbolTable.getCurrentFunction());
		}
		return new StringType();
   }

   @Override public IType visitExprToBool(pdrawParser.ExprToBoolContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
      if (exprType != Type.BOOLEAN && exprType != Type.INTEGER && exprType != Type.REAL) {
         ErrorHandler.error("Conversion to boolean can only be applied to boolean values.", ctx, symbolTable.getCurrentFunction());
      }

      return new BooleanType();
   }

   @Override public IType visitExprToInt(pdrawParser.ExprToIntContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
		if (!isConvertibleToNumeric(exprType)) {
			ErrorHandler.error("Conversion to integer can only be applied to real, integer, or string values.", ctx, symbolTable.getCurrentFunction());
		}
		return new IntegerType();
   }

   @Override public IType visitExprMultDivMod(pdrawParser.ExprMultDivModContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
		Type rightType = visit(ctx.expression(1)).getType();
		if (!isNumericType(leftType) || !isNumericType(rightType)) {
			ErrorHandler.error("Operands of '*' or '/' must be numeric.", ctx, symbolTable.getCurrentFunction());
      }
		if (ctx.op.getText().equals("//") || ctx.op.getText().equals("\\\\")) {
			return new IntegerType();
		} else if (leftType == Type.REAL || rightType == Type.REAL) {
			return new Real();
      } else if (ctx.op.getText().equals("/")){
         return new Real();
		}

		return new IntegerType();
   }

   @Override public IType visitExprAddSub(pdrawParser.ExprAddSubContext ctx) {
		Type leftType = visit(ctx.expression(0)).getType();
		Type rightType = visit(ctx.expression(1)).getType();
		if (!((isNumericType(leftType) && isNumericType(rightType)) || (leftType == Type.PEN && rightType == Type.POINT) || (leftType == Type.POINT && rightType == Type.POINT))) {
			ErrorHandler.error("Operands of '+' or '-' must be numeric.", ctx, symbolTable.getCurrentFunction());
		}
      return (leftType == Type.PEN ? new Pen() : leftType == Type.POINT ? new Point() : (leftType == Type.REAL || rightType == Type.REAL) ? new Real() : new IntegerType());
   }

   @Override public IType visitExprSetProperty(pdrawParser.ExprSetPropertyContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
		Type rightType = visit(ctx.expression(1)).getType();

		if (leftType != Type.PEN) {
			ErrorHandler.error("Set property operator can only be applied to pen values.", ctx, symbolTable.getCurrentFunction());
		}

		String property = ctx.Property().getText();
		switch (property) {
			case "color":
				if (rightType != Type.INTEGER) {
					ErrorHandler.error("Color value must be a integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "pressure":
				if (!isNumericType(rightType)) {
					ErrorHandler.error("Pressure value must be a real or integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "thickness":
				if (rightType != Type.INTEGER) {
					ErrorHandler.error("Thickness value must be a integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;	
			case "orientation":
				if (!isNumericType(rightType)) {
					ErrorHandler.error("Orientation value must be a real or integer value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
			case "position":
				if (rightType != Type.POINT) {
					ErrorHandler.error("Position value must be a point value.", ctx, symbolTable.getCurrentFunction());
				}
				break;
		
			default:
				ErrorHandler.error("Invalid property " + property, ctx, symbolTable.getCurrentFunction());
				break;

		}
		return new Pen();
   }

   @Override public IType visitExprInteger(pdrawParser.ExprIntegerContext ctx) {
      return new IntegerType();
   }

   @Override public IType visitExprId(pdrawParser.ExprIdContext ctx) {
      String id = ctx.ID().getText();
      Symbol s = symbolTable.getSymbol(id);
      if (s == null) {
         ErrorHandler.error(id + " was not declared.", ctx, symbolTable.getCurrentFunction());

         return null;
      }

      return s.getIType();
   }

   @Override public IType visitExprString(pdrawParser.ExprStringContext ctx) {
      return new StringType();
   }

   @Override public IType visitExprParent(pdrawParser.ExprParentContext ctx) {
      return visit(ctx.expression());
   }

   @Override public IType visitExprComp(pdrawParser.ExprCompContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
      Type rightType = visit(ctx.expression(1)).getType();
      return new BooleanType();
   }

   @Override public IType visitExprPoint(pdrawParser.ExprPointContext ctx) {
      return new Point();
   }

   @Override public IType visitExprStringConcat(pdrawParser.ExprStringConcatContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
      Type rightType = visit(ctx.expression(1)).getType();
      if (leftType != Type.STRING || rightType != Type.STRING) {
         ErrorHandler.error("Operands of '++' must be strings.", ctx, symbolTable.getCurrentFunction());
      }
      return new StringType();
   }

   @Override public IType visitExprColor(pdrawParser.ExprColorContext ctx) {
      return new IntegerType();
   }

   @Override public IType visitExprToReal(pdrawParser.ExprToRealContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
      if (!isConvertibleToNumeric(exprType)) {
         ErrorHandler.error("Conversion to real can only be applied to real, integer, or string values.", ctx, symbolTable.getCurrentFunction());
      }
      return new Real();
   }

   @Override public IType visitExprPenOperator(pdrawParser.ExprPenOperatorContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
      Type rightType = visit(ctx.expression(1)).getType();
      if (leftType != Type.PEN || !isNumericType(rightType)) {
         ErrorHandler.error("Operands of pen operator must be pen values.", ctx, symbolTable.getCurrentFunction());
      }
      return new Pen();
   }

   @Override public IType visitExprConvToRad(pdrawParser.ExprConvToRadContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
      if (!isConvertibleToNumeric(exprType)) {
         ErrorHandler.error("Conversion to radian can only be applied to real, integer, or string values.", ctx, symbolTable.getCurrentFunction());
      }
      return new Real();
   }

   @Override public IType visitExprPenUnary(pdrawParser.ExprPenUnaryContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
      if (exprType != Type.PEN) {
         ErrorHandler.error("Unary operator can only be applied to pen values.", ctx, symbolTable.getCurrentFunction());
      }
      return new Pen();
   }

   @Override public IType visitExprUnary(pdrawParser.ExprUnaryContext ctx) {
      IType exprType = visit(ctx.expression());
      if (!isNumericType(exprType.getType())) {
         ErrorHandler.error("Unary operator can only be applied to numeric values.", ctx, symbolTable.getCurrentFunction());
      }
      return exprType;
   }

   @Override public IType visitExprNew(pdrawParser.ExprNewContext ctx) {
      if (ctx.expression() != null) {
         Type exprType = visit(ctx.expression()).getType();
         if (exprType != Type.PENTYPE) {
            ErrorHandler.error("New operator can only be applied to pen type values.", ctx, symbolTable.getCurrentFunction());
         }
      }
      return new Pen();
   }

   @Override public IType visitExprBoolOp(pdrawParser.ExprBoolOpContext ctx) {
      Type leftType = visit(ctx.expression(0)).getType();
      Type rightType = visit(ctx.expression(1)).getType();
      if (leftType != Type.BOOLEAN || rightType != Type.BOOLEAN) {
         ErrorHandler.error("Operands of boolean operators must be boolean values.", ctx, symbolTable.getCurrentFunction());
      }
      return new BooleanType();
   }

   @Override public IType visitExprStdin(pdrawParser.ExprStdinContext ctx) {
      Type exprType = visit(ctx.expression()).getType();
      if (exprType != Type.STRING) {
         ErrorHandler.error("Stdin operator can only be applied to string values.", ctx, symbolTable.getCurrentFunction());
      }
      return new StringType();
   }

   @Override public IType visitExprAssign(pdrawParser.ExprAssignContext ctx) {
      return visit(ctx.assign());
   }

   @Override public IType visitAssign(pdrawParser.AssignContext ctx) {
      String id = ctx.ID().getText();
      Symbol symbol = symbolTable.getSymbol(id);
      if (symbol == null) {
         ErrorHandler.warning("Variable with id " + id + " was not declared.", ctx, symbolTable.getCurrentFunction());

         // Implicit declaration
         symbol = new Symbol(id, new IntegerType(), ctx.start.getLine(), ctx.start.getCharPositionInLine(), ctx.stop.getLine(), ctx.stop.getCharPositionInLine());
         symbolTable.addSymbol(symbol);
      }
      IType exprType = visit(ctx.expression());
      if (!hasImplicitConvertion(exprType.getType(), symbol.getIType().getType())) {
         ErrorHandler.error("Type mismatch in assignment. Cannot assign " + exprType.getType() + " to " + symbol.getIType().getType() + ".", ctx, symbolTable.getCurrentFunction());
         if (symbol.getCodeLocation() != null)
            ErrorHandler.info("Symbol `\033[1m" + id + "\033[0m` was declared at line " + symbol.getCodeLocation().getStartLine() + " column " + symbol.getCodeLocation().getStartColumn(), ctx, symbolTable.getCurrentFunction(), symbol);
      }
      return exprType;
   }

   @Override public IType visitExprReal(pdrawParser.ExprRealContext ctx) {
      return new Real();
   }

   @Override public IType visitExprBool(pdrawParser.ExprBoolContext ctx) {
      return new BooleanType();
   }

   @Override public IType visitPoint(pdrawParser.PointContext ctx) {
      Type exprType1 = visit(ctx.expression(0)).getType();
      Type exprType2 = visit(ctx.expression(1)).getType();
      if (!isNumericType(exprType1) || !isNumericType(exprType2)) {
         ErrorHandler.error("Point values must be integer or real values.", ctx, symbolTable.getCurrentFunction());
      }
      return new Point();
   }

   @Override public IType visitReturn(pdrawParser.ReturnContext ctx) {
      Symbol function = symbolTable.getCurrentFunction();
      if (function == null) {
         ErrorHandler.error("Return statement must be inside a function.", ctx, symbolTable.getCurrentFunction());
      }

      IType returnType = function.getReturnType();
      IType exprType = visit(ctx.expression());
      if (!hasImplicitConvertion(exprType.getType(), returnType.getType())) {
         ErrorHandler.error("Return type " + exprType.getType() + " does not match function type " + returnType.getType() + ".", ctx, symbolTable.getCurrentFunction());
      }
      return null;
   }

   @Override public IType visitFunction(pdrawParser.FunctionContext ctx) {
      String id = ctx.ID().getText();
      IType returnType = switchFunctionType(ctx.type.getText());

      ParameterList parameterSymbols;
      if (ctx.parameter_list() != null)
         parameterSymbols = (ParameterList) visit(ctx.parameter_list());
      else
         parameterSymbols = new ParameterList(); // empty parameter list

      IType type = new FunctionType(returnType, parameterSymbols);
      Symbol function = new Symbol(id, type, ctx.start.getLine(), ctx.start.getCharPositionInLine(), ctx.stop.getLine(), ctx.stop.getCharPositionInLine());
      symbolTable.addSymbol(function);

      symbolTable.enterScope(function);
      for (Symbol parameterSymbol : parameterSymbols.getParameterSymbols()) {
         symbolTable.addSymbol(parameterSymbol);
      }

      visit(ctx.statement());
      symbolTable.exitScope();
      return null;
   }

   @Override public IType visitParameter(pdrawParser.ParameterContext ctx) {
      ctx.id = ctx.ID().getText();
      IType type = switchFunctionType(ctx.type.getText());
      
      return type;
   }

   @Override public IType visitParameter_list(pdrawParser.Parameter_listContext ctx){
      ParameterList res = new ParameterList();

      for (pdrawParser.ParameterContext parameter : ctx.parameter()) {
         IType parType = visit(parameter);
         res.add(new Symbol(parameter.id, parType));
      }

      return res;
   }

   @Override public IType visitExprFunctionCall(pdrawParser.ExprFunctionCallContext ctx){
      String id = ctx.ID().getText();
      Symbol function = symbolTable.getSymbol(id);
      if (function == null) {
         ErrorHandler.error("Function with id " + id + " does not exist.", ctx, symbolTable.getCurrentFunction());
      }

      FunctionType functionType = (FunctionType) function.getIType();
      ParameterList parameterLists = functionType.getParameterList();
      if (ctx.expression().size() != parameterLists.getParameterSymbols().size()) {
         ErrorHandler.error("Function call " + id + " has wrong number of arguments.", ctx, symbolTable.getCurrentFunction());
      }

      for (int i = 0; i < ctx.expression().size(); i++) {
         IType exprType = visit(ctx.expression(i));
         IType parameterList = parameterLists.getParameterSymbols().get(i).getIType();
         if (!hasImplicitConvertion(exprType.getType(), parameterList.getType())) {
            ErrorHandler.error("Function call " + id + " has wrong argument type.", ctx, symbolTable.getCurrentFunction());
         }
      }

      return functionType.getReturnType();
   }

   @Override public IType visitExprPi(pdrawParser.ExprPiContext ctx) {
      return new Real();
   }

   // private function to see if expr is INTEGER or REAL
	private boolean isNumericType(Type type) {
		return type == Type.INTEGER || type == Type.REAL;
	}
  
	// types that can be converted besides string
	private boolean isConvertibleToString(Type type) {
		return type == Type.INTEGER || type == Type.REAL || type == Type.STRING;
	}

	// types that can be converted to numeric
	private boolean isConvertibleToNumeric(Type type) {
		return type == Type.INTEGER || type == Type.REAL || type == Type.STRING;
	}

   private boolean hasImplicitConvertion(Type from, Type to) {
      return from == to || (from == Type.INTEGER && to == Type.REAL);
   }
   
}

File: src/pen/semanticVerifier.py
from antlr4 import *
from penParser import penParser
from penVisitor import penVisitor
from vTypes import Type

class SemanticVerifier(penVisitor):
   def __init__(self, errorHandler):
      self.variables = {} # {variable_name: type}
      self.errorHandler = errorHandler
      
   def switchFunctionType(self, vType):
      if vType == 'int':
         return Type.Integer
      elif vType == 'real':
         return Type.Real
      elif vType == 'string':
         return Type.String
      elif vType == 'bool':
         return Type.Bool
      elif vType == 'point':
         return Type.Point
      else:
         return None
      

   def visitProgram(self, ctx:penParser.ProgramContext):
      self.visitChildren(ctx)
      return None

   def visitDeclaration_element(self, ctx:penParser.Declaration_elementContext):
      return None
   
   def visitCompound(self, ctx:penParser.CompoundContext):
      return self.visitChildren(ctx)
   
   def visitBlock(self, ctx:penParser.BlockContext):
      return self.visitChildren(ctx)

   def visitDeclaration(self, ctx:penParser.DeclarationContext):
      vType = self.switchFunctionType(ctx.type_.getText())
      for element in ctx.declaration_element():
         if element.ID():
            var_name = element.ID().getText()
            self.variables[var_name] = vType
         elif element.assign():
            var_name = element.assign().ID().getText()
            self.variables[var_name] = vType
            self.visit(element.assign())
      return None

   def visitStatement(self, ctx:penParser.StatementContext):
      return self.visitChildren(ctx)

   def visitWhile(self, ctx:penParser.WhileContext):
      if (self.visit(ctx.expression()) != Type.Bool):
         self.errorHandler.error("While condition must be a boolean value.", ctx.start.line, ctx.start.column)
      self.visit(ctx.statement())
      return None

   def visitUntil(self, ctx:penParser.UntilContext):
      if (self.visit(ctx.expression()) != Type.Bool):
         self.errorHandler.error("Until condition must be a boolean value.", ctx.start.line, ctx.start.column)
      self.visit(ctx.statement())
      return None

   def visitFor(self, ctx:penParser.ForContext):
      self.visit(ctx.statement(0))
      if (self.visit(ctx.expression(0)) != Type.Bool):
         self.errorHandler.error("Loop range must be a boolean value.", ctx.start.line, ctx.start.column)
      self.visit(ctx.expression(1))
      self.visit(ctx.statement(1))
      return None

   def visitIf(self, ctx:penParser.IfContext):
      if (self.visit(ctx.expression()) != Type.Bool):
         self.errorHandler.error("If condition must be a boolean value.", ctx.start.line, ctx.start.column)
      self.visit(ctx.statement(0))
      if ctx.statement(1):
         self.visit(ctx.statement(1))
      return None

   def visitPause(self, ctx:penParser.PauseContext):
      if (self.visit(ctx.expression()) != Type.Integer):
         self.errorHandler.error("Pause time must be an integer value.", ctx.start.line, ctx.start.column)
      return None

   def visitStdout(self, ctx:penParser.StdoutContext):
      exprType = self.visit(ctx.expression())
      if (exprType != Type.String and not self.isNumericType(exprType) and exprType != Type.Bool):
         self.errorHandler.error("Stdout value must be a string.", ctx.start.line, ctx.start.column)
      return None

   def visitExprToString(self, ctx:penParser.ExprToStringContext):
      exprType = self.visit(ctx.expression())
      if not (self.isNumericType(exprType) or self.isStringType(exprType) or self.isPointType(exprType) or exprType == Type.Bool):
         self.errorHandler.error("Conversion to string can only be applied to real, integer, bool, point or string values.", ctx.start.line, ctx.start.column)
      return Type.String

   def visitExprToBool(self, ctx:penParser.ExprToBoolContext):
      exprType = self.visit(ctx.expression())
      if (exprType != Type.Bool and exprType != Type.Integer and exprType != Type.Real and exprType != Type.String):
         self.errorHandler.error("Conversion to boolean can only be applied to real, integer, or boolean values.", ctx.start.line, ctx.start.column)
      return Type.Bool

   def visitExprToInt(self, ctx:penParser.ExprToIntContext):
      exprType = self.visit(ctx.expression())
      if not (self.isNumericType(exprType) or self.isStringType(exprType)):
         self.errorHandler.error("Conversion to integer can only be applied to real, integer, or string values.", ctx.start.line, ctx.start.column)
      return Type.Integer
   
   def visitExprMultDivMod(self, ctx:penParser.ExprMultDivModContext):
      exprType1, exprType2 = self.visit(ctx.expression(0)), self.visit(ctx.expression(1))
      if not (self.isNumericType(exprType1) and self.isNumericType(exprType2)):
         self.errorHandler.error("The values must be numeric (Integer or Real)", ctx.start.line, ctx.start.column)
      result = Type.Integer
      if (exprType1 == Type.Real or exprType2 == Type.Real):
         result = Type.Real
      if ctx.op.text == '//':
         result = Type.Integer
      if ctx.op.text == '/':
         result = Type.Real
      return result

   def visitSetProperty(self, ctx:penParser.SetPropertyContext):
      prop = ctx.Property().getText()
      if prop == 'color':
         if (self.visit(ctx.expression()) != Type.Integer):
            self.errorHandler.error("Color value must be an integer.", ctx.start.line, ctx.start.column)
      elif prop == 'pressure':
         if not self.isNumericType(self.visit(ctx.expression())):
            self.errorHandler.error("Pressure value must be a number.", ctx.start.line, ctx.start.column)
      elif prop == 'thickness':
         if not self.isNumericType(self.visit(ctx.expression())):
            self.errorHandler.error("Thickness value must be a number.", ctx.start.line, ctx.start.column)
      elif prop == 'orientation':
         if not self.isNumericType(self.visit(ctx.expression())):
            self.errorHandler.error("Orientation value must be a numver.", ctx.start.line, ctx.start.column)
      elif prop == 'position':
         if self.visit(ctx.expression()) != Type.Point:
            self.errorHandler.error("Position value must be a point.", ctx.start.line, ctx.start.column)
      return None

   def visitExprAddSub(self, ctx:penParser.ExprAddSubContext):
      exprType1, exprType2 = self.visit(ctx.expression(0)), self.visit(ctx.expression(1))
      if not (self.isNumericType(exprType1) and self.isNumericType(exprType2)):
         self.errorHandler.error("The values must be numeric (Integer or Real)", ctx.start.line, ctx.start.column)
      result = Type.Integer
      if (exprType1 == Type.Real or exprType2 == Type.Real):
         result = Type.Real
      return result

   def visitExprInteger(self, ctx:penParser.ExprIntegerContext):
      return Type.Integer

   def visitExprId(self, ctx:penParser.ExprIdContext):
      id = ctx.ID().getText()
      if id not in self.variables:
         self.errorHandler.error("Variable " + id + " not declared.", ctx.start.line, ctx.start.column)
      return self.variables.get(id)

   def visitExprPi(self, ctx:penParser.ExprPiContext):
      return Type.Real

   def visitExprString(self, ctx:penParser.ExprStringContext):
      return Type.String

   def visitExprParent(self, ctx:penParser.ExprParentContext):
      return self.visit(ctx.expression())

   def visitExprComp(self, ctx:penParser.ExprCompContext):
      exprType1, exprType2 = self.visit(ctx.expression(0)), self.visit(ctx.expression(1))
      if not (self.isNumericType(exprType1) and self.isNumericType(exprType2) 
          or exprType1 == exprType2):
         self.errorHandler.error("Type mismatch in comparison. Cannot compare " + str(exprType1) + " with " + str(exprType2) + ".", ctx.start.line, ctx.start.column)
      return Type.Bool
         
   def visitExprPoint(self, ctx:penParser.ExprPointContext):
      return Type.Point

   def visitExprStringConcat(self, ctx:penParser.ExprStringConcatContext):
      exprType1 = self.visit(ctx.expression(0))
      exprType2 = self.visit(ctx.expression(1))
      if not (self.isStringType(exprType1) and self.isStringType(exprType2)):
         self.errorHandler.error("Operands of concatenation must be strings.", ctx.start.line, ctx.start.column)
      return Type.String

   def visitExprColor(self, ctx:penParser.ExprColorContext):
      return Type.Integer

   def visitExprToReal(self, ctx:penParser.ExprToRealContext):
      exprType = self.visit(ctx.expression())
      if not (self.isNumericType(exprType) or self.isStringType(exprType)):
         self.errorHandler.error("Conversion to real can only be applied to real, integer, or string values.", ctx.start.line, ctx.start.column)
      return Type.Real

   def visitPenOperator(self, ctx:penParser.PenOperatorContext):
      if not self.isNumericType(self.visit(ctx.expression())):
         self.errorHandler.error("Pen operator can only be applied to real or integer values.", ctx.start.line, ctx.start.column)
      return None

   def visitExprConvToRad(self, ctx:penParser.ExprConvToRadContext):
      return Type.Real

   def visitPenUnary(self, ctx:penParser.PenUnaryContext):
      return None

   def visitExprUnary(self, ctx:penParser.ExprUnaryContext):
      exprType = self.visit(ctx.expression())
      if not (self.isNumericType(exprType)):
         self.errorHandler.error("Unary operator can only be applied to numeric values.", ctx.start.line, ctx.start.column)
      return exprType

   def visitExprBoolOp(self, ctx:penParser.ExprBoolOpContext):
      exprType1 = self.visit(ctx.expression(0))
      exprType2 = self.visit(ctx.expression(1))
      if exprType1 != Type.Bool or exprType2 != Type.Bool:
         self.errorHandler.error("Operands of boolean operators must be boolean values.", ctx.start.line, ctx.start.column)
      return Type.Bool

   def visitExprStdin(self, ctx:penParser.ExprStdinContext):
      exprType = self.visit(ctx.expression())
      if not (self.isStringType(exprType)):
         self.errorHandler.error("Stdin operator can only be applied to string values.", ctx.start.line, ctx.start.column)
      return Type.String

   def visitExprAssign(self, ctx:penParser.ExprAssignContext):
      return self.visit(ctx.assign())

   def visitExprReal(self, ctx:penParser.ExprRealContext):
      return Type.Real

   def visitExprBool(self, ctx:penParser.ExprBoolContext):
      return Type.Bool

   def visitAssign(self, ctx:penParser.AssignContext):
      id = ctx.ID().getText()
      exprType = self.visit(ctx.expression())
      if id not in self.variables:
         self.errorHandler.warning("Variable `" + id + "` not declared.", ctx.start.line, ctx.start.column)
         self.variables[id] = exprType
         
      if not self.hasImplicitConversion(exprType, self.variables[id]):
         self.errorHandler.error("Type mismatch in assignment. Cannot assign " + str(exprType) + " to " + str(self.variables[id]) + ".", ctx.start.line, ctx.start.column)
      return self.variables[id]

   def visitPoint(self, ctx:penParser.PointContext):
      exprType1 = self.visit(ctx.expression(0))
      exprType2 = self.visit(ctx.expression(1))
      if not (self.isNumericType(exprType1) and self.isNumericType(exprType2)):
           self.errorHandler.error("Point values must be integer or real values.", ctx.start.line, ctx.start.column)
      return Type.Point
   
   def isNumericType(self, vType):
      return vType == Type.Integer or vType == Type.Real
   
   def isStringType(self, vType):
      return vType == Type.String
   
   def isPointType(self, vType):
      return vType == Type.Point
   
   def hasImplicitConversion(self, _from, _to):
      return _from == _to or (_from == Type.Integer and _to == Type.Real)
   